### PR TITLE
ARROW-9775: [C++] Automatic S3 region selection

### PR DESCRIPTION
**Pull request by pitrou**
Once looked up, the region of a given bucket is cached for the lifetime of the process.
(the cache size is unbounded, which shouldn't be a problem in practice)

S3Options::FromUri() now resolves the region for the bucket if it wasn't given as a Uri query parameter.

### CODE REVIEW

**Comment by pitrou** [comment]
@nealrichardson

File: cpp/src/arrow/util/atomic_shared_ptr.h [hunk 1]
```diff
@@ -89,5 +89,25 @@ inline void atomic_store(
   *p = r;
 }
 
+template <class T>
```
Comment (pitrou) [inline cpp/src/arrow/util/atomic_shared_ptr.h:92]: @bkietz No concern here?

File: cpp/src/arrow/type_traits.h [hunk 2]
```diff
@@ -22,7 +22,7 @@
 #include <type_traits>
 #include <vector>
 
-#include "arrow/type_fwd.h"
+#include "arrow/type.h"
```
Comment (bkietz) [inline cpp/src/arrow/type_traits.h:25]: this change seems odd; was type_fwd incorrect?

File: cpp/src/arrow/util/atomic_shared_ptr.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/util/atomic_shared_ptr.h:92]: Looks correct to me

File: cpp/src/arrow/util/atomic_shared_ptr.h [hunk 3]
```diff
@@ -89,5 +89,25 @@ inline void atomic_store(
...
+inline bool atomic_compare_exchange_strong(
```
Comment (bkietz) [inline cpp/src/arrow/util/atomic_shared_ptr.h:93]: templates are implicitly `inline`
```suggestion
bool atomic_compare_exchange_strong(
```

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/util/atomic_shared_ptr.h (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/util/atomic_shared_ptr.h:93]: All other functions are explicitly marked inline in this file, but I can make the change.

File: cpp/src/arrow/type_traits.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/type_traits.h:25]: Yes, there are errors such as:
```
../src/arrow/type_traits.h:149:1: error: incomplete type 'arrow::UInt8Type' named in nested name specifier
PRIMITIVE_TYPE_TRAITS_DEF(uint8_t, UInt8, uint8)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/arrow/type_traits.h:145:14: note: expanded from macro 'PRIMITIVE_TYPE_TRAITS_DEF'
      CType, ARROW_CONCAT(ArrowShort, Type), ARROW_CONCAT(ArrowShort, Array), \
      ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/arrow/util/macros.h:24:28: note: expanded from macro 'ARROW_CONCAT'
#define ARROW_CONCAT(x, y) x##y
                           ^
<scratch space>:126:1: note: expanded from here
UInt8Type
^
../src/arrow/type_traits.h:130:19: note: expanded from macro 'PRIMITIVE_TYPE_TRAITS_DEF_'
    using CType = ArrowType_::c_type;                                                    \
                  ^~~~~~~~~~~~
../src/arrow/type_fwd.h:186:1: note: forward declaration of 'arrow::UInt8Type'
_NUMERIC_TYPE_DECL(UInt8)
^
../src/arrow/type_fwd.h:176:9: note: expanded from macro '_NUMERIC_TYPE_DECL'
  class KLASS##Type;                                  \
        ^
<scratch space>:274:1: note: expanded from here
UInt8Type
^
```

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
Better wait for #8218 before merging this.

**Pull request closed without merging**